### DISCUSSION TITLE
Coupling PySWMM with MODFLOW

### DISCUSSION THREAD

**Discussion by vprigiob** [pyswmm/pyswmm#364] [Q&A] [answered]
Hi,
I would like to interact with the Module Aquifer in SWMM to update the aquifer properties.
I write for example:
from pyswmm import Simulation, Aquifers
pyswmm gives me an error.

Is it possible to update the aquifer properties through pyswmm? or this feature is not available yet?
I cannot find any example about this.

Thank you,
VP

**Comment by abhiramm7** [comment]
Hi @vprigiob, 

As far as I know, pyswmm does not have this feature yet. Currently, we support updating the features listed in https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/subcatchments.py . 

Best, 
Abhi

**Comment by vprigiob** [comment]
Hi Abhi,

Thanks for getting back.

I am working on the coupling of  swmm with a model that describes the groundwater and
I need to update the aquifer properties such as the water table into swmm, continuously.

It would be great if this feature could be added at pyswmm.

The project is a USGS funded project for flooding prediction where we are coupling open-source
hydrology models (such as modflow, swmm, fflow) using python.

Is it possible to work together to add missing feature in pyswmm to allow the coupling?

Thank you.

Best regards,
Valentina Prigiobbe
----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Reply by abhiramm7** [reply]
Hi Dr. Prigiobbe,

Yeah, absolutely.

SWMM has had issues representing groundwater phenomena; I love the idea of having the ability to interface SWMM with other groundwater models. 

I would need more specifics about the parameters in SWMM that we want to interact with modflow to give you a specific list of steps that would have to be done. But on a high level, we would have to do the following:

1. Update SWMM source code to expose the groundwater model parameters
2. Add the python function to update groundwater parameters from pyswmm

Of these tasks, 1 is the most time-consuming.

Please feel to reach out if you have any questions :) 

Best, 
Abhi

**Reply by dickinsonre** [reply]
SWMM is not a GW model - it is only intended to provide some underground infiltration routing.  It has a limited goal for the surficial aquifer, IMO.  It is open source so linking to other models is always welcome,

**Reply by abhiramm7** [reply]
@dickinsonre do you think it might be a good option to swap out the entire module?

**Comment by vprigiob** [comment]
Hi Abhiram,

The second option (Add the python function to update groundwater parameters from pyswmm) is better because we want to use modflow for the groundwater and swmm for the sewer modeling.
In SWMM, it is possible to determine the infiltration of groundwater into the pipes.
But the groundwater level needs to be provided.
This is done by creating an aquifer for a subcatchment and entering the water table elevation of that aquifer.

What I would like to do using pyswmm is to update the water table elevation, which we calculate with modflow.
It would be also great if all the aquifer properties could be updated through pyswmm.

Thanks,
Valentina Prigiobbe
----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Reply by bemcdonnell** [reply]
@vprigiob,  This sounds pretty interesting.  I am supportive of exposing the parameters you need to couple MODFLOW.   UDM last year included a nice talk by Teresita Scheuch (https://udm2022.org/wp-content/uploads/2021/08/UDM2022_Extended_Abstract-4d2f431b.pdf) on this subject.  My plan was to follow up with Teresita after the conference to see what sorts of limitations she ran into with pyswmm.   

At the moment, you have access to HGL inside the network (as I am sure you are aware).  Considering @dickinsonre's comment in the other thread, the GW process is super limited in SWMM.  Would it be easier / cleaner to expose subcatchment infiltration rates and, using MODFLOW, could you configure it to represent the water elevation, receiving infiltration, and exfil to the sewer?  On the PYSWMM side, you are able to pythonically "inject flow" during a running simulation to a node using the `node.generated_inflow()` method.  So essentially, if you know the HGL differential between the soil and the network, you could come up with a method to calibrate the I&I(HGL Diff).  

This is just an idea!  I want to acknowledge my naivety of MODFLOW since I've never used it 😇.  My hope is to simplify what needs to be exposed so we can streamline the feature request.  Moreover, I hope we can invite you to collaborate on the development!

**Reply by abhiramm7** [reply]
@vprigiob yeah for sure. I like the idea. There are multiple ways we can approach this integration. Let me know if you want to hop on a call to work out the logistics.

**Reply by tscheuch** [reply]
Thanks for the mention @bemcdonnell.
@vprigiob as Bryant explained before, we built an integration scheme to couple SWMM and MODFLOW using the libraries PySWMM and FloPy. A bidirectional exchange of fluxes between SWMM (i.e., infiltration) and MODFLOW (i.e., drained water) was proposed. Thus simulation results of the SWMM model have an impact on the MODFLOW model and vice versa. This aproach is necessary in urban settings with shallow groundwater.

At the moment we are working on python library to be able to apply this scheme in different case studies. I will keep you updated with the progress. I hope to have news soon!

If you want to know more about integration scheme feel free to contact me.

**Reply by bemcdonnell** [reply]
@tscheuch, is your project made available in an open source repository? As I understand, @vprigiob is looking to do a simple example (to later be expanded on). Might be very timely with your current research!

**Comment by vprigiob** [comment]
Hi Abhim,
Yes, we can have a call on Zoom.
I am located on the east coast nearby New York and tomorrow morning anytime is good for me.
Please let me know if tomorrow morning is good for you otherwise we can find another time.
Thanks,
Valentina


----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Reply by abhiramm7** [reply]
Works for me, I'll send out a meeting invite.

**Comment by vprigiob** [comment]
Ok, great.
Thanks,
Valentina

----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Comment by vprigiob** [comment]
Hi,

I am interested to learn more about the code for the exchange. Is it a python package that it is possible to download?
I am working on the same thinks and I would be happy to compare the approaches.

Thanks,
Valentina

----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Comment by vprigiob** [comment]
Hi,
Mu project is not in a repository yet as we are still working on it.
I am actually @vprigiob<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fvprigiob&data=05%7C01%7Cvprigiob%40stevens.edu%7C0877ce2817f247bac77b08daa183387e%7C8d1a69ec03b54345ae21dad112f5fb4f%7C0%7C0%7C637999882223994698%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=dYDKismRVhLGpIRo4K%2FwdgvxJRiYIGgeiCOJRPnjogw%3D&reserved=0> .

Best,
Valentina

----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Comment by vprigiob** [comment]
Hi Abhiram,

We had a conversation few weeks ago about the input of infiltration to a swmm model to create an interaction between swmm and modflow.
I am now working on the coupling of the two software and I would like to run swimm for given time intervals nested within the modflow model.
For instance, the goal is to run the groundwater model for a week and at the end of this time calculate the flux exchanges for the sewer and I run swmm.
I know how the two software can interact, but I cannot find in the pyswmm tutorials and examples a case where the simulation of a swmm model is time controlled, paused, and then re-started again.

Please let me know if there is any example I may have missed. I can share the code with you with some more details if necessary.

Thank you in advance.

Best regards,
Valentina Prigiobbe

----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Comment by abhiramm7** [comment]
Hi Dr. Prigiobbe, 

There are a few examples on how you can use pyswmm for running the kind of simulation you are describing. I have couple of follow-up questions to make sure I understand your ask. 

1. When you say pause, do you mean pausing the simulation in python i.e., .out and .rpt are not created and you are haulting the solver after a couple of timesteps? 
2. Will you be altering your .inp file when you pause your simulation?

If your answer is **_yes_** on the first and **_no_** on the second, [this example](https://gist.github.com/abhiramm7/3d50199cd58862bd10a3bd9f70b496c1) might be helpful. If not, we might have to hack together a solution :). 

I have added Bryant to the thread, he might have more insights. 

Best, 
Abhi

CC: @bemcdonnell

**Comment by bemcdonnell** [comment, marked as answer]
@vprigiob, looks like some good progress so far!  So it sounds like you’re looking to ask MODFLOW to control the time step.  I think this is probably a good approach.  Is there a Python Library available to help with this?  If you want MODFLOW to control the step, this means that the method in which you would run Pyswmm is subtly different from the conventional method that we promote for most pyswmm users. The workflow is generally as follows:

1. Open and initialize your MODFLOW model. I am assuming you have some Python wrapper to this and you have the ability to discreetly control MODFLOW as you do in Pyswmm for SWMM?
2. Open your Pyswmm Simulation. 
3. step your MODFLOW model
4. get the time step and HGLs from MODFLOW
5. Pass these to PySWMM 
6. Configure the governing equations to drive inflow into the PySWMM model
7. Advance your simulation forward using pyswmm's simulation iterator
8. repeat step 3 Until simulations are complete
9. clean exit/close MODFLOW and PySWMM

I have attached some example code and a ZIP file with a test model.  
[ExampleModels.zip](https://github.com/OpenWaterAnalytics/pyswmm/files/9990636/ExampleModels.zip)



```python
import pyswmm

# Open and Initialize MODFLOW model
mod_flow_model = 0 # Initial Time Step

# Open and Initialize the SWMM Simulation
sim = pyswmm.Simulation("Example1.inp")

node_of_interest = pyswmm.Nodes(sim)["21"]

# Start PySWMM Simulation
sim.start()

while True:

    if sim._terminate_request == True:
        break

    # Run MODFLOW for the desired time step and return dt
    delta_t = 900 # seconds (This was returned from MODFLOW)
    GW_head = 991 # ft (this is returned from MODFLOW)

    # There are some nuances about this method but for the
    # purposes of longer-term simulation of I&I, they are
    # probably not mission critical to understand
    sim.step_advance(delta_t)

    # Code Required to End your Simulation Cleanly
    try:
        sim.next() # seconds
    except StopIteration:
        break

    # Get the Head from the Junction!
    head = node_of_interest.head

    # Some Infiltration Rate Governing Equation
    infil_rate = (head - GW_head) * 80
    # Filter negative
    if infil_rate < 0:
        infil_rate = 0

    node_of_interest.generated_inflow(infil_rate)

sim.__exit__()

```

if you run my example you will see inside the RPT file that we are adding a major inflow to Junction 21. 

```
  *******************
  Node Inflow Summary
  *******************
  
  -------------------------------------------------------------------------------------------------
                                  Maximum  Maximum                  Lateral       Total        Flow
                                  Lateral    Total  Time of Max      Inflow      Inflow     Balance
                                   Inflow   Inflow   Occurrence      Volume      Volume       Error
  Node                 Type           CFS      CFS  days hr:min    10^6 gal    10^6 gal     Percent
  -------------------------------------------------------------------------------------------------
  9                    JUNCTION      4.66     4.66     0  04:01       0.402       0.402       0.000
  10                   JUNCTION      4.52     9.17     0  04:01       0.388        0.79      -0.000
  13                   JUNCTION      2.45     2.45     0  04:01       0.202       0.202       0.000
  14                   JUNCTION      0.00     2.44     0  04:01           0       0.202       0.000
  15                   JUNCTION      6.56     9.00     0  04:01        0.57       0.771       0.000
  16                   JUNCTION      0.00    32.99     0  04:01           0        23.2       0.000
  17                   JUNCTION      0.00    31.46     0  04:13           0        23.2       0.000
  19                   JUNCTION      0.79     0.79     0  04:01      0.0558      0.0558       0.000
  20                   JUNCTION      0.00     0.78     0  04:02           0      0.0558       0.000
  21                   JUNCTION    160.00   165.42     0  04:02         148         149       0.020
  22                   JUNCTION      2.45    31.86     0  04:01       0.202        27.6       0.004
  23                   JUNCTION      1.50     1.50     0  04:01       0.124       0.124       0.000
  24                   JUNCTION      0.00    33.05     0  04:01           0        23.3      -0.000
  18                   OUTFALL       1.33    32.31     0  04:14       0.108        23.3       0.000
```

**Comment by vprigiob** [comment]
Hi Bryant,

Thank you for providing the example. I have tried it to implement but I am not sure swmm is running correctly.

I use Python ModflowApi package to communicate with modflow.
Currently I am running modflow until convergence and the I use the calculated head to determine the infiltration/exfiltration for swmm.
Infiltration/exfiltration will be also considered sink/source terms for modflow.
I have followed your example and I have implemented in the following way:


```python
simsw.start()
while True:
    if simsw._terminate_request == True:
        break
    else:
        # MODFLOW -----------------------------------------------
        while current_time<end_time:
        #HERE MODFLOW RUNS UNTIL CONVERGENCE AND UNTIL THE END OF current_time. THE CALCULATED 
        #HEAD IS USED TO DETERMINE Infiltration.
        # SWMM -----------------------------------------------
            J7.generated_inflow(infiltration)  # CFS
            simsw.step_advance(t15d)
            out_array[j] = np.array(Out1.total_inflow) # one value of the flow rate is saved within the time period
            try:
                simsw.next(current_time*3600*24) # seconds, current_time is the time the modflow sim ends
            except StopIteration:
                break
try:
        mf6.finalize()
        success = True
except:
        raise RuntimeError
simsw._exit_()
```


I am not sure sure the integration of the two sorfate is correct. I know that the code works for modflow and I get correct results.
But when I add swmm there are two main errors I receive:

  *   for simsw.next(). Should I write the end of the time interval to finish the swmm simulation temporally and pass the control to modflow again?
  *   for simsw._exit_(). It seems that there is not the attribute "exit" and I am wondering if the syntax is different.

If I comment out  simsw._exit_(), the code runs but there is not flow rate output. So, I believe I am not calling swmm properly.
Thank you in advance for your help.

Best,
Valentina

----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041

**Reply by bemcdonnell** [reply]
@vprigiob, the simsw.next() does not take any arguments. When you declare the sim.step_advance(seconds) that’s where you decide the step length

**Reply by bemcdonnell** [reply]
@vprigiob also, what version of python/pyswmm are you using?

**Comment by vprigiob** [comment]
Hi Bryant,

Ok, perfect. There was a comment (# seconds) that confused me.
So I provide a step advance equal to the duration of the modflow simulation, which is 15 days.
After this I believe  swmm is in pause and it re-starts again from where it was left using as initial condition the last simulation results from the previous run. Am I right?

I am using Python 3.9 and pyswmm 1.2

Thank you so much,
Valentina
----------------------------------
Valentina Prigiobbe, Ph.D.
Assistant Professor
Department of Civil, Environmental, and Ocean Engineering
Stevens Institute of Technology
Hoboken NJ, USA
we<http://personal.stevens.edu/~vprigiob/>bsite<http://personal.stevens.edu/~vprigiob/>
Zoom link: https://stevens.zoom.us/j/8641218041
________________________________

**Reply by bemcdonnell** [reply]
@vprigiob i think the better way to say is that swmm pauses, then continues.  We are not actually stopping the model, it is just waiting for your process to complete before it steps forward.

**Reply by bemcdonnell** [reply]
@vprigiob, 

I edited your comment to adjust the code you posted with python formatting.  Can you confirm this is what your code looks like? 

https://github.com/OpenWaterAnalytics/pyswmm/discussions/364

Also, it's a lot easier if you log into github and reply rather than use the email response feature.  If my edits to the code that you posted are correct, we might need to make some subtle adjustments.

**Reply by vprigiob** [reply]
Hi Bryant,

Thanks for the editing. I have updated the code and changed some indents, as well.
The results are now good and they are similar to those I obtain running for the swmm GUI just with a different time step.

Thank you very much for you help.

Regards,
Valentina

**Reply by bemcdonnell** [reply]
@vprigiob I'm so glad to hear! Keep us posted on your progress!
